lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+
+  // But just knowing which registers we need is not sufficient, we also have
+  // to order them properly.
+  auto Ordered = abi::orderArguments(ArgumentRegisters, ABI);
+
+  llvm::SmallVector<model::Argument, 8> Result;
+  for (model::Register::Values Register : Ordered) {
+    auto CurrentRegisterIterator = Used.find(Register);
+    if (CurrentRegisterIterator != Used.end()) {
+      // If the current register is confirmed to be in use, convert it into
+      // an argument while preserving its type and name.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    } else {
+      // This register is unused but we still have to create an argument
+      // for it. Otherwise the resulting function will not be compatible with
+      // the current one: if the "hole" is not preserved, there will be no way
+      // to recreate it at any point in the future, when, for example,
+      // converting back to the original-like representation.
+      auto &Current = Result.emplace_back();
+      if (Used.at(Register).Type().UnqualifiedType().get() != nullptr)
+        Current.Type() = Used.at(Register).Type();
+      else
+        Current.Type() = { defaultRegisterType(Register), {} };
+      Current.CustomName() = Used.at(Register).CustomName();
+    }
+  }
+
+  // Rewrite all the indices to make sure they are consistent.
+  for (auto Pair : llvm::enumerate(llvm::reverse(Result)))
+    Pair.value().Index() = Pair.index();
+
+  return Result;
+}
+
+std::optional<llvm::SmallVector<model::Argument, 8>>
+TDManager::tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                                       size_t IndexOffset) const {
+  // Qualifiers here are not allowed since this must point to a simple struct.
+  revng_assert(StackArgumentTypes.Qualifiers().empty());
+
+  auto *Unqualified = StackArgumentTypes.UnqualifiedType().get();
+  if (!Unqualified) {
+    // If there is no type, it means that the importer responsible for this
+    // function didn't detect any stack arguments and avoided creating
+    // a new empty type.
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
+  revng_assert(Pointer != nullptr,
+               "`RawFunctionType::StackArgumentsType` must be a struct");
+  const model::StructType &Types = *Pointer;
+
+  // If the struct is empty, it indicates that there are no stack arguments.
+  if (Types.Fields().empty()) {
+    revng_assert(Types.size() == 0);
+    return llvm::SmallVector<model::Argument, 8>{};
+  }
+
+  // Compute the alignment of the first argument.
+  uint64_t CurrentAlignment = *ABI.alignment(Types.Fields().begin()->Type());
```

Comment:
* Please in the doxygen of this method explain what's the assumption we're making about the the state of the stack arguments `struct`. I guess we basically expect one field per argument, is that right?
* I see at least three instance of `Types.Fields().begin()`, make a local variable.

Reply:
1. Is this what you had in mind?
https://github.com/revng/revng/pull/232/commits/69ea94581907a60a16d20cc27856de5d73aad87f
```
  /// This function expects the struct to be strictly formatted:
  ///   - it has to have the field at offset 0;
  ///   - any field after the first has to be placed at an offset that can be
  ///     predicted by looking at its alignment and the sizes of all the fields
  ///     that precede it.
  ///   - the same condition applies to the relation between the last field and
  ///     the total size of the struct.
  ///
  /// If any field's location derives from the predicted one, the struct is
  /// deemed non-ABI-conformant and the conversion is aborted.
  ///
  /// \note for invalid type and empty structs (with size 0), an empty vector
  ///       is returned. This is NOT a failure, just the fact that function does
  ///       not accept any arguments using the stack.
  ///
```

2. Done.

Reply:
Gorgeous